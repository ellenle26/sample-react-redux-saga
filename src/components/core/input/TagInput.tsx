import classNames from "classnames";
import { FC, useRef } from "react";

const TagInput: FC<TagInputProps> = (props: TagInputProps) => {
    const required = props.required === true;
    const requiredMargin = required ? "mr-2" : "";
    const border = props.error ? "border-error" : "border-border";

    const inputRef = useRef<HTMLInputElement>(null);


    return (
        <div className={classNames("w-full", { "opacity-50": props.disabled }, props.className)}>
            {(props.title || required) && (
                <div className="mb-1 flex items-center text-sm font-bold">
                    {props.title && (
                        <p className={`${requiredMargin}`}>{props.title}</p>
                    )}
                </div>
            )}
            <div
                className={`h-fit flex items-start flex-wrap w-full bg-white text-gray-800 text-base box-border border outline-none ${border} pt-2 px-4 rounded-lg`}
                onClick={() => {
                    inputRef.current?.focus();
                }}
            >
                {
                    Array.isArray(props.values) && props.values.length > 0 && props.values.map((t,i) => (
                        <div key={i} className="min-w-fit flex items-center mr-2 mb-2 bg-white border border-tag rounded-2xl px-2 py-1">
                            <span className="leading-none mr-1">{t}</span>
                            <img className="cursor-pointer mt-1" alt="close_icon" src="/icons/close.svg" 
                                onClick={() => props.onRemoveTag && props.onRemoveTag(i)}/>
                        </div>
                    ))
                }
                <input
                    ref={inputRef}
                    className="outline-none text-sm mt-1 mb-2"
                    type="text"
                    onKeyDown={(event) => {
                        if (event.key === "Enter") {
                            if (event.currentTarget.value && event.currentTarget.value.trim() !== '') {
                                props.onAddTag && props.onAddTag(event.currentTarget.value);
                                event.currentTarget.value = ''
                            }
                        }
                    }}
                />
            </div>
            {props.error && (
                <p className="text-xs mt-1 text-error">{props.error}</p>
            )}
        </div>
    )
}

export default TagInput;

interface TagInputProps {
    title?: string;
    name?: string;
    placeholder?: string;
    required?: boolean;
    values: string[];
    error?: string;
    disabled?: boolean;
    className?: string;
    onAddTag: (value: string) => void;
    onRemoveTag: (id: number) => void;
}