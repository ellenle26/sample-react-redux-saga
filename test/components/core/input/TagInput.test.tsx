/**
 * @jest-environment jsdom
*/

import TagInput from "@components/core/input/TagInput";
import { fireEvent, render } from "@testing-library/react";
import React from "react";


const values = ["a", "b", "c", "d"];
const add = jest.fn();
const remove = jest.fn();

test("snapshot testing", () => {
    const renderResult = render(<TagInput  values={values} onAddTag={add} onRemoveTag={remove}/>);
    expect(renderResult.asFragment()).toMatchSnapshot();
});

test("render items testing", () => {
    const renderResult = render(<TagInput values={values} onAddTag={add} onRemoveTag={remove}/>);
    expect(renderResult.container.children[0].children.length).toEqual(1);
    // expect 4 tags and 1 input
    expect(renderResult.container.children[0].children[0].children.length).toEqual(5);
});

test("remove tag testing", () => {
    const renderResult = render(<TagInput values={values} onAddTag={add} onRemoveTag={remove}/>);
    const tags = renderResult.container.getElementsByClassName("min-w-fit flex items-center mr-2 mb-2 bg-white border border-tag rounded-2xl px-2 py-1")
    fireEvent.click(tags[0].getElementsByTagName("img")[0]);
    expect(remove).toHaveBeenCalledTimes(1);
});


test("add tag testing", () => {
    const renderResult = render(<TagInput values={values} onAddTag={add} onRemoveTag={remove}/>);
    const input = renderResult.container.getElementsByTagName("input")[0];
    fireEvent.change(input, { target: { value: "test" } });
    fireEvent.keyDown(input, { key: "Enter" });
    expect(add).toHaveBeenCalledTimes(1);
})

test("add empty tag testing", () => {
    const renderResult = render(<TagInput values={values} onAddTag={add} onRemoveTag={remove} />);
    const input = renderResult.container.getElementsByTagName("input")[0];
    fireEvent.change(input, { target: { value: " " } });
    fireEvent.keyDown(input, { key: "Enter" });
    expect(add).toHaveBeenCalledTimes(1);
})

test("different key down testing", () => {
    const renderResult = render(<TagInput values={values} onAddTag={add} onRemoveTag={remove} />);
    const input = renderResult.container.getElementsByTagName("input")[0];
    fireEvent.change(input, { target: { value: " " } });
    fireEvent.keyDown(input, { key: "Shift" });
    expect(add).toHaveBeenCalledTimes(1);
})

test('err and required testing', () => {
    const renderResult = render(<TagInput values={values} onAddTag={add} onRemoveTag={remove} error="error" required title="test"/>);
    const ptag = renderResult.container.getElementsByClassName("text-xs mt-1 text-error")[0];
    expect(ptag).toBeTruthy();
})