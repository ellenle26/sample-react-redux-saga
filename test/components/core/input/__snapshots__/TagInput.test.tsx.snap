// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot testing 1`] = `
<DocumentFragment>
  <div
    class="w-full"
  >
    <div
      class="h-fit flex items-start flex-wrap w-full bg-white text-gray-800 text-base box-border border outline-none border-border pt-2 px-4 rounded-lg"
    >
      <div
        class="min-w-fit flex items-center mr-2 mb-2 bg-white border border-tag rounded-2xl px-2 py-1"
      >
        <span
          class="leading-none mr-1"
        >
          a
        </span>
        <img
          alt="close_icon"
          class="cursor-pointer mt-1"
          src="/icons/close.svg"
        />
      </div>
      <div
        class="min-w-fit flex items-center mr-2 mb-2 bg-white border border-tag rounded-2xl px-2 py-1"
      >
        <span
          class="leading-none mr-1"
        >
          b
        </span>
        <img
          alt="close_icon"
          class="cursor-pointer mt-1"
          src="/icons/close.svg"
        />
      </div>
      <div
        class="min-w-fit flex items-center mr-2 mb-2 bg-white border border-tag rounded-2xl px-2 py-1"
      >
        <span
          class="leading-none mr-1"
        >
          c
        </span>
        <img
          alt="close_icon"
          class="cursor-pointer mt-1"
          src="/icons/close.svg"
        />
      </div>
      <div
        class="min-w-fit flex items-center mr-2 mb-2 bg-white border border-tag rounded-2xl px-2 py-1"
      >
        <span
          class="leading-none mr-1"
        >
          d
        </span>
        <img
          alt="close_icon"
          class="cursor-pointer mt-1"
          src="/icons/close.svg"
        />
      </div>
      <input
        class="outline-none text-sm mt-1 mb-2"
        type="text"
      />
    </div>
  </div>
</DocumentFragment>
`;
